### PR TITLE
Add more tests for the SPDX-License-Identifier

### PR DESCRIPTION
**Pull request by axic**
Part of #10145.

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `98c3d050ac00a379fdb08b34e06fefb9e361aa5c`:
```
Error: Trailing whitespace found:
 test/libsolidity/syntaxTests/license/license_whitespace_2.sol:1:// SPDX-License-Identifier: GPL-3.0 

```
Please check that your changes are working as intended.

**Comment by axic** [comment]
> Please check that your changes are working as intended.

They actually are. This is what the test is about.

File: test/libsolidity/syntaxTests/license/license_hidden_unicode.sol
Comment (axic) [inline test/libsolidity/syntaxTests/license/license_hidden_unicode.sol:2]: Basically can't avoid creating metadata tests to compare what goes into the output 😢

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `ad1ae8e9435f9e970c87b05f676c76a7a6589268`:
```
Error: Trailing whitespace found:
 test/libsolidity/syntaxTests/license/license_whitespace_2.sol:1:// SPDX-License-Identifier: GPL-3.0 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `5afd637f015c64f58015ebb743a186cdb0a51153`:
```
Error: Trailing whitespace found:
 test/libsolidity/Metadata.cpp:363: // SPDX-License-Identifier: GPL-3.0 

```
Please check that your changes are working as intended.

**Comment by axic** [comment]
Come on @stackenbotten.

Is there a way to put some file on an exception list?

File: test/libsolidity/Metadata.cpp
Comment (cameel) [inline test/libsolidity/Metadata.cpp]: You could silence @stackenbotten by ditching the raw string in this particular test:
```suggestion
	char const* sourceCode =
		"// NOTE: we also add trailing whitespace after the license, to see it is trimmed.\n"
		"// SPDX-License-Identifier: GPL-3.0    \n"
		"\"pragma solidity >=0.0;\"\n"
		"contract C {\n"
		"}";
```

**Comment by cameel** [comment]
@axic There actually is [a spot in `check_style.sh` that we're using as an ad-hoc whitelist](https://github.com/ethereum/solidity/blob/v0.8.0/scripts/check_style.sh#L9-L12):

``` bash
WHITESPACE=$(git grep -n -I -E "^.*[[:space:]]+$" |
  grep -v "test/libsolidity/ASTJSON\|test/libsolidity/ASTRecoveryTests\|test/compilationTests/zeppelin/LICENSE" |
  grep -v -E "test/libsolidity/syntaxTests/comments/unicode_direction_override_1.sol"
)
```

But I added a suggestion above that I think is better than disabling this check for the whole file.

File: test/libsolidity/Metadata.cpp
Comment (axic) [reply test/libsolidity/Metadata.cpp]: Good idea!

**Comment by axic** [comment]
@cameel thanks, took both suggestions!

File: test/libsolidity/Metadata.cpp
Comment (cameel) [inline test/libsolidity/Metadata.cpp]: What do you think about putting the repeating part in a helper?
Or we could have a helper to validate the relevant part of the returned JSON (i.e. that the `sources` key is there, that there's only one source, etc.) and another that returns a list of licenses found in the metadata or `nullopt` so that we can do comparison here.

File: test/libsolidity/syntaxTests/license/license_missing.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/license/license_missing.sol:1]: Since you added metadata tests for this, it should be removed, right?

File: test/libsolidity/syntaxTests/license/license_whitespace_1.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/license/license_whitespace_1.sol]: One more possible test case:
```solidity
// SPDX-License-Identifier    : GPL-3.0
contract C {}
```

File: test/libsolidity/syntaxTests/license/license_whitespace_3.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/license/license_whitespace_3.sol]: I'd add a comment below saying that there's trailing whitespace here. Some people might have editors configured no to show it (github does not show it either) and just trim it without realizing.

I'd also rename it to `license_trailing_whitespace.sol`.

File: test/libsolidity/syntaxTests/license/license_whitespace_1.sol
Comment (cameel) [reply test/libsolidity/syntaxTests/license/license_whitespace_1.sol]: Also more test ideas:
- Whole comment inside Unicode direction override marks (`/// <LRO>SPDX-License-Identifier: GPL-3.0<PDF>`)
- Missing colon (`/// SPDX-License-Identifier GPL-3.0`)
- Multi-line NatSpec comment (`/** SPDX-License-Identifier: GPL-3.0 */`)
- Multi-line comment (`/* SPDX-License-Identifier: GPL-3.0 */`)
- License comment at the bottom of the file
- License comment ending with a Mac or Windows newline
- License comment on a non-empty line (`pragma solidity ^0.8.0; // SPDX-License-Identifier: GPL-3.0`)
- License comment in a string (`import "// SPDX-License-Identifier: GPL-3.0"`)
- No whitespace at all (`//SPDX-License-Identifier:GPL-3.0`, `/*SPDX-License-Identifier:GPL-3.0*/`)

**Review by cameel — COMMENTED** [review]
The logic change for preamble is very simple and seems fine. I only have some suggestions on improving the tests.

File: test/libsolidity/Metadata.cpp
Comment (cameel) [inline test/libsolidity/Metadata.cpp]: I'd give the file a name. Less guessing what the `""` is referring to below.
```suggestion
	compilerStack.setSources({{"A.sol", std::string(sourceCode)}});
```

**Comment by cameel** [comment]
Looks like some tests are failing because not adding the license line changed the source locations where errors happen. Fortunately only 5 are failing (I'd expect many more). They probably contain the license comment already and I think it's best to just remove it.

File: test/libsolidity/Metadata.cpp
Comment (axic) [reply test/libsolidity/Metadata.cpp]: Honestly I don't want to spend more time on this PR than neccesary. We should move standardjson/metadata tests into a similar system the other ones work. I think spending time on that compared to writing helpers in the existing boost tests may be better.

Comment (cameel) [reply test/libsolidity/Metadata.cpp]: Good point. It's fine then.

File: test/libsolidity/syntaxTests/license/license_missing.sol
Comment (axic) [reply test/libsolidity/syntaxTests/license/license_missing.sol:1]: The reason I kept is I hope the metadata test is temporary.

Comment (cameel) [reply test/libsolidity/syntaxTests/license/license_missing.sol:1]: OK. I'd put that it the comment (perhaps even referencing the metadata test and this file from metadata test) but now I understand the intention.

File: test/libsolidity/syntaxTests/license/license_whitespace_1.sol
Comment (axic) [reply test/libsolidity/syntaxTests/license/license_whitespace_1.sol]: If you have the time please feel free to add them.

Comment (cameel) [reply test/libsolidity/syntaxTests/license/license_whitespace_1.sol]: Not today unfortunately and I'll be off since Monday (though I'll probably work a bit over the weekend just to finally finish the stuff with binaries and later to get them reviewed if anyone is still available to do it). If you're fine putting this on hold until I'm back, feel free to assign the PR to me. Otherwise I can approve the PR even without them once everything else in the PR works (tests are failing currently). The tests you added already add some value so there's no reason not to merge it just because we can add even more :)

File: test/libsolidity/SyntaxTest.cpp
Comment (bshastry) [inline test/libsolidity/SyntaxTest.cpp:60]: It is possible that the license identifier is not added in corner cases like this

```
contract C {
// SPDX-License-Identifier: GPL-3.0
}
```

**Comment by axic** [comment]
> Looks like some tests are failing because not adding the license line changed the source locations where errors happen. Fortunately only 5 are failing (I'd expect many more). 

The longer it will take to merge this, the more times we'll need to adjust tests :wink:

**Comment by leonardoalt** [comment]
@axic @cameel if you have time we could try to fix and merge this at once

**Comment by cameel** [comment]
OK. I'll take it over and add the extra tests I suggested.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `d5cf5989f65792fee9a44fda4e53dd8f07fc015c`:
```
Error: Trailing whitespace found:
 test/libsolidity/syntaxTests/license/license_cr_endings.sol:1:// This test is actually useless, as the test suite adds the automatic preamble.^M// SPDX-License-Identifier: GPL-3.0^Mcontract C {}^M
 test/libsolidity/syntaxTests/license/license_crlf_endings.sol:1:// This test is actually useless, as the test suite adds the automatic preamble.^M
 test/libsolidity/syntaxTests/license/license_crlf_endings.sol:2:// SPDX-License-Identifier: GPL-3.0^M
 test/libsolidity/syntaxTests/license/license_crlf_endings.sol:3:contract C {}^M
 test/libsolidity/syntaxTests/license/license_whitespace_trailing.sol:1:// SPDX-License-Identifier: GPL-3.0 

```
Please check that your changes are working as intended.

File: test/libsolidity/SyntaxTest.cpp
Comment (cameel) [reply test/libsolidity/SyntaxTest.cpp:60]: Yeah, it's not added in this case. I added this as a test case.

File: test/libsolidity/syntaxTests/license/license_whitespace_1.sol
Comment (cameel) [reply test/libsolidity/syntaxTests/license/license_whitespace_1.sol]: I have now added all these test cases. Both as `.sol` files and as metadata tests to see if the license is actually picked up.

There actually was a bug in the code and `import "// SPDX-License-Identifier: GPL-3.0"` detected it - `SyntaxTest.cpp` not only adds the preamble but also subtracts the added characters from error locations. This second part was still assuming the old preamble which resulted in wrong locations. I initially thought that it's a bug on `develop` in the code that reports the error but after some debugging I realized that it's only code from this PR that messes things up.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `7186752f761f2ead8a6efbcbbca81a7a04e3e4e2`:
```
Error: Trailing whitespace found:
 test/libsolidity/syntaxTests/license/license_cr_endings.sol:1:// This test is actually useless, as the test suite adds the automatic preamble.^M// SPDX-License-Identifier: GPL-3.0^Mcontract C {}^M
 test/libsolidity/syntaxTests/license/license_crlf_endings.sol:1:// This test is actually useless, as the test suite adds the automatic preamble.^M
 test/libsolidity/syntaxTests/license/license_crlf_endings.sol:2:// SPDX-License-Identifier: GPL-3.0^M
 test/libsolidity/syntaxTests/license/license_crlf_endings.sol:3:contract C {}^M
 test/libsolidity/syntaxTests/license/license_whitespace_trailing.sol:1:// SPDX-License-Identifier: GPL-3.0 

```
Please check that your changes are working as intended.

**Comment by cameel** [comment]
This is now finished and ready for another review.

All my changes to @axic's commits (except for those required to rebase properly) are left in unsquashed fixups for now and here's the summary of the bigger ones:
- `SyntaxTest::filterObtainedErrors()` was not updated to subtract the new preamble from source locations in error messages. This is the bug I was talking about earlier. But it was not on `develop` after all.
- Added the extra tests cases I recommended. Both as `*.sol` files and as metadata tests.
- Refactored duplicate code out of metadata license tests.

Please take a look at the results of metadata tests. In some cases they are different than I expected. I'll point them out in code comments.

File: test/libsolidity/Metadata.cpp
Comment (cameel) [inline test/libsolidity/Metadata.cpp:487]: This one is interesting (suggested by @bshastry). The license is detected when defined before or after the contract but not inside it. It this a bug or a feature? :)

File: test/libsolidity/Metadata.cpp
Comment (cameel) [inline test/libsolidity/Metadata.cpp:514]: So I was wrong in https://github.com/ethereum/solidity/issues/10145#issuecomment-720199176 that licenses Natspec comments are not recognized? Has it always been like this or is this a recent change?

File: test/libsolidity/Metadata.cpp
Comment (cameel) [inline test/libsolidity/Metadata.cpp:523]: Interesting that even multi-line comments work.

File: test/libsolidity/Metadata.cpp
Comment (cameel) [inline test/libsolidity/Metadata.cpp:550]: And comments in random places like this.

File: test/libsolidity/syntaxTests/license/license_bottom.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/license/license_bottom.sol:3]: I have added `*.sol` equivalents for the metadata tests, just like @axic originally did (see https://github.com/ethereum/solidity/pull/10653#discussion_r545974629), but there are a lot of them and they are indeed useless, so maybe better to remove them after all?

**Comment by cameel** [comment]
Looks like `t_bytecode_compare` is failing because of #10771: the Windows build uses author date while others use committer date and in this particular PR these dates are different. I created #10791 to fix that.

File: test/libsolidity/Metadata.cpp
Comment (bshastry) [reply test/libsolidity/Metadata.cpp:487]: Sadly, license identifier is not part of our antlr grammar spec yet. I'm guessing this is a feature i.e., global license declarations only.

File: test/libsolidity/syntaxTests/license/license_bottom.sol
Comment (bshastry) [reply test/libsolidity/syntaxTests/license/license_bottom.sol:3]: The fuzzer copies tests from the `test` directory. So perhaps retain them?

**Review by bshastry — DISMISSED** [review]
LGTM if tests pass

File: test/libsolidity/syntaxTests/license/license_bottom.sol
Comment (cameel) [reply test/libsolidity/syntaxTests/license/license_bottom.sol:3]: OK. That sounds like a good enough reason to leave them in.

**Comment by chriseth** [comment]
Needs squashing.

**Comment by cameel** [comment]
I think it will pass because the only failing test was the bytecode comparison that was broken on `develop` for a moment.

Before merging I'll still need to squash the fixups.

File: test/libsolidity/SyntaxTest.cpp
Comment (chriseth) [inline test/libsolidity/SyntaxTest.cpp]: This is removed if `NDEBUG` is defined. Do we really want to use c assert?

Comment (cameel) [reply test/libsolidity/SyntaxTest.cpp]: I think the C assert a good thing because it lets me not worry if the check is too heavy or if I'm doing too many of them. I'd say that if we're worrying that it might actually happen, then it's not really an assertion and should be properly validated instead.

In Solidity I'm usually using `solAssert()` though because that's the prevailing style and I want to respect that, but I checked other files in this directory to see which variant to use (`solAssert()`, `assertThrow()`, etc.) and the few uses I saw were actually `assert()` so I went with it.

I can replace it with `solAssert()` if you prefer.

Comment (cameel) [reply test/libsolidity/SyntaxTest.cpp]: I went ahead and changed them to `solAssert()` now.

**Comment by cameel** [comment]
Fixups squashed. Let's make sure all tests pass before merging - I didn't have time to rerun local tests after replacing the assert yet.

**Review by bshastry — APPROVED** [review]
LGTM, reapproving

**Pull request merged**
